src/k8s-engine/client.ts

```diff
@@ -166,38 +194,43 @@ export class KubeClient extends events.EventEmitter {
     this.forwarder = new k8s.PortForward(this.kubeconfig, true);
     this.shutdown = false;
     this.coreV1API = this.kubeconfig.makeApiClient(k8s.CoreV1Api);
+    this.services = null;
+  }
 
-    // Set up a watch for services
-    // Since the watch API we have _doesn't_ notify us when things have
-    // changed, we'll need to do some trickery and wrap the underlying watcher
-    // with our own code.
-    class WrappedWatch extends k8s.Watch {
-      callback: ()=>void;
-      constructor(kubeconfig: k8s.KubeConfig, callback:()=>void) {
-        super(kubeconfig);
-        this.callback = callback;
-      }
+  async waitForServiceWatcher() {
+    const startTime = Date.now();
+    const maxWaitTime = 300_000;
+    const waitTime = 3_000;
 
-      watch(
-        path: string,
-        queryParams: any,
-        callback: (phase: string, apiObj: any, watchObj?: any) => void,
-        done: () => void,
-        error: (err: any) => void
-      ): Promise<any> {
-        const wrappedCallback = (phase: string, apiObj: any, watchObj?: any) => {
-          callback(phase, apiObj, watchObj);
-          this.callback();
-        };
+    while (true) {
+      const currentTime = Date.now();
 
-        return super.watch(path, queryParams, wrappedCallback, done, error);
+      if ((currentTime - startTime) > maxWaitTime) {
+        console.log(`Waited more than ${ maxWaitTime / 1000 } secs, it might start up later`);
+        break;
       }
+      if (await this.getServiceListWatch()) {
+        break;
+      }
+      await util.promisify(setTimeout)(waitTime);
     }
+  }
 
+  async getServiceListWatch() {
+    if (this.services) {
+      return this.services;
+    }
+    if ((await this.coreV1API.listServiceForAllNamespaces()).body.items.length === 0) {
```

Comment:
Please add a comment on why this is okay (because there are Kubernetes-internal services, so it should never be empty).

Reply:
It'sthe result of an experiment. Stepping into the code leads into what looks like a comment-free elegant state machine that sits at the top of a complex beast. Sticking with my observation that trying to understand why this works is an unnecessary rabbit hole.

Reply:
Ah, I'm not asking for how it under the hood; just a comment on why this is will always succeed even for clusters with no user-created services (because the cluster has internal services).

Reply:
I don't know why we get zero services here, just that we do during the early stages of k8s starting up.

Reply:
It would still good to say _something_ around this line, so that we can understand it in a month…

Reply:
added a comment